scheduler/starter/modules.go

```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		return nil, err
+	}
+	protocolFactory := thrift.NewTBinaryProtocolFactoryDefault()
+
+	wf := func(node cluster.Node) runner.Service {
+		di := dialer.NewSimpleDialer(transportFactory, protocolFactory, common.DefaultClientTimeout)
+		cl, _ := client.NewSimpleClient(di, string(node.Id()))
+		return runners.NewPollingService(cl, cl, 250*time.Millisecond)
+	}
+
+	return wf, nil
+}
+
+func GetProxyAddr() string {
+	port := os.Getenv("SOCKS_PORT")
+	if port != "" {
+		return "localhost:" + port
+	}
+	return ""
+}
+
+func GetProxyDialer(proxyAddr string) (proxy.Dialer, error) {
+	return proxy.SOCKS5("tcp", proxyAddr, nil, proxy.Direct)
+}
+
+func NewTTransportFactory() (thrift.TTransportFactory, error) {
+	addr := GetProxyAddr()
+	if addr == "" {
+		addr = DEFAULT_SOCKS_ADDR
+	}
+	dialer, err := GetProxyDialer(addr)
+	if err != nil {
+		return nil, err
+	}
+	return &socksTransportFactory{dialer}, nil
+}
+
+type socksTransportFactory struct {
+	dialer proxy.Dialer
+}
+
+func (f *socksTransportFactory) GetTransport(trans thrift.TTransport) thrift.TTransport {
+	// If this is a SOCKS-able connection, then return a TTransport that uses SOCKS.
+	// If it's not, return trans unmodified.
+	tsock, ok := trans.(*thrift.TSocket)
+	if !ok {
+		// Don't know how to wrap anything other than a TSocket
+		return trans
+	}
+	if tsock.IsOpen() {
+		// Transport Factory should be called before Open
+		return trans
+	}
+	addr, ok := tsock.Addr().(*net.TCPAddr)
+	if !ok {
+		return trans
+	}
+	ip := addr.IP.To4()
+	if ip == nil {
+		return trans
+	}
+
+	// An Address is Socks-able if it starts with 10, because Twitter's production network lives in
+	// 10/8 in IPv4 space.
+	if ip[0] != 10 {
+		return trans
+	}
+
+	if f.dialer == nil {
+		// We want to use SOCKS, but can't, because we have no dialer
+		log.Info("Aurora address specified, but cannot use SOCKS proxy because SOCKS_PORT is unset", addr.String())
+		return trans
+	}
+
+	return &socksSocket{
+		conn:    nil,
+		addr:    addr.String(),
+		timeout: 0, // thrift.TSocket gives us no way to extract the timeout
+		dialer:  f.dialer,
+	}
+}
+
+// socksSocket creates a socket via the proxy dialer.
+// (Ideally, there would be a form of TSocket that took a Dialer instead of a net.Conn)
+type socksSocket struct {
+	conn    *thrift.TSocket
+	addr    string
+	timeout time.Duration
+	dialer  proxy.Dialer
+}
+
+func (s *socksSocket) IsOpen() bool {
+	return s.conn != nil
+}
+
+func (s *socksSocket) Open() error {
+	if s.IsOpen() {
+		return s.conn.Open()
+	}
+	conn, err := s.dialer.Dial("tcp", s.addr)
+	if err != nil {
+		log.Infof("Failed connecting to %s via proxy, trying directly\n", s.addr)
+		conn, err = net.DialTimeout("tcp", s.addr, time.Duration(10*time.Second))
+		if err != nil {
+			return err
+		}
+	}
+
+	s.conn = thrift.NewTSocketFromConnTimeout(conn, 0)
+	return nil
+}
+
+func (s *socksSocket) Read(buf []byte) (int, error) {
+	return s.conn.Read(buf)
```

Comment:
should we embed conn as an unnamed field? means we can skip defining these funcs

Reply:
see prior comment

Reply:
not sure i follow... if we had the socksSocket struct defined like this

`type socksSocket struct {
	*thrift.TSocket
	addr    string
	timeout time.Duration
	dialer  proxy.Dialer
}`

we could remove Read, Write, Close, Flush, & RemainingBytes from here, as socksSocket.Read(buf) would just delegate to the *thrift.TSocket